src/s3path.jl

```diff
@@ -71,15 +85,18 @@ function S3Path(
 end
 
 # To avoid a breaking change.
-function S3Path(str::AbstractString; config::AWSConfig=global_aws_config())
+function S3Path(str::AbstractString; config::AbstractAWSConfig=global_aws_config())
     result = tryparse(S3Path, str; config=config)
     result !== nothing || throw(ArgumentError("Invalid s3 path string: $str"))
     return result
 end
 
-function Base.tryparse(::Type{S3Path}, str::AbstractString; config::AWSConfig=global_aws_config())
+# if config=nothing, will not try to talk to AWS until after string is confirmed to be an s3 path
+function Base.tryparse(::Type{<:S3Path}, str::AbstractString; config::Union{Nothing,AbstractAWSConfig}=nothing)
```

Comment:
Can this not be?

> ; config::AbstractAWSConfig=global_aws_config())

Then remove the `(config === nothing)`

Reply:
No, that's the whole point.

What was happening here was that any time you try to parse a path, that is *ANY* path with `p"/a/b/c"`, because of the way `FilePathsBase.register` works the *very first thing* that would happen (if you don't have AWS environment variables) is that it starts trying to talk to the AWS API to try to get EC2 credentials.

So for example, I was testing this on one of my machines with min.io.  I have no AWS environment variables and am not an AWS's environment.  So, if I do `p"/etc/fstab"` (yes, even a filesystem path) it hangs for a long time and then ultimately errors because it called `global_aws_config()`, did not have local environment variables, and could not talk to the AWS end point where it would get EC2 credentials.  Arguably, this would be bad even on AWS because an unexpectedly expensive network call is taking place even when you are doing something as innocent as declaring a file system path.

The work around you see here partially solves the problem by ensuring that `global_aws_config()` is *NOT* called unless the path it is trying to parse really is an S3 path.

Again, I personally find even this a little dubious, and if it were up to me I'd remove `FilePathsBase.register` from the initialization of AWSS3 (users who wanted it would still have the option of adding it themselves).  However, the changes I have made above are non-breaking, so there's no need to decide whether you agree with me on that right now.

Reply:
Ah okay, thank you for clarifying! This solution definitely makes sense to go with, and also should lead to some expanse in our unit testing to run a some suite w/o credentials so we can keep track of these things.

Reply:
Great, thanks.  Yeah, some more tests are probably needed.  I'm not going to offer to add any since it's so hard for me to test this package.

Btw, after this is merged, I'm going to publish Minio.jl, which is just a very small package that makes it convenient to declare a config for minio, and also makes it convenient to fetch and run the minio binary as a server.  Part of why I set that up is to make testing easier, and testing this package could be a possible application of this.

Reply:
Logged an issue for it:

- https://github.com/JuliaCloud/AWSS3.jl/issues/142

I'll get around to it at some point